Glossary and terms (#1264)
fishtown-analytics/docs.getdbt.com: website/docs/terms/subquery.md

Subqueries are usually contrasted with <Term id="cte">Common Table Expressions (CTEs)</Term> as they have similar use cases. Unlike CTEs, which are usually separate `SELECT` statements within a query, subqueries are usually `SELECT` statements nested within a `JOIN`, `FROM`, or `WHERE` statement in a query.

To be honest, we rarely write subqueries here at dbt Labs since we prefer to use CTEs. We find that CTEs, in general, support better query readability, organization, and debugging. However, subqueries are a foundational concept in SQL and still widely used. We hope you can use this glossary to better understand how to use subqueries and how they differ from CTEs.

## Subquery syntax

While there are technically several types of subqueries, the general syntax to build them is the same. A subquery usually consists of the following:

- Enclosing parentheses
- A name
- An actual SELECT statement
- A main query it is nested in via a FROM, WHERE, or JOIN clause

Let’s take this to an example, using the [sample jaffle_shop dataset](https://github.com/dbt-labs/jaffle_shop).

```sql
select customer_id, count(order_id) as cnt_orders
  from (

>         select * from {{ ref(‘orders’) }}

Reviewer: orders is an actual string:

```sql
select customer_id, count(order_id) as cnt_orders
  from (
        select * from {{ ref('orders') }}
       ) all_orders
group by 1
```

       ) all_orders
group by 1
```

Given the elements of subqueries laid out in the beginning, let’s break down this example into its respective parts.

| Subquery elements | Example |
|---|---|
| Enclosing parentheses | :white_check_mark: |
| Subquery name | `all_orders` |

> | `SELECT` statement | `select * from {{ ref(`orders`) }}` | 

Reviewer: I think we want to do this here:  `select * from {{ ref('orders') }}`

Reviewer 2: Okay thanks

| Main query it is nested in | `select customer_id, count(order_id) as cnt_orders from all_orders group by 1` | 

> When this query is actually executed, it will start by running the innermost query first. In this case, it would run `select * from {{ ref(‘orders’) }}` first (NOTE: If you want to learn more about what a `ref`, [check out our documentation on it](https://docs.getdbt.com/reference/dbt-jinja-functions/ref). Then, it would pass those results to the outer query, which is where you grab the count of orders by `customer_id`.

Reviewer: need a closing parenthesis for NOTE

Also want to have it be `select * from {{ ref('orders') }}`

Reviewer 2: Going to make this a standalone note block to make it more apparent.

In a correlated subquery, the outer query will execute row-by-row. For each row, that result from the outer query will be passed to the inner query. Compare this to nested queries: in a nested query, the inner query is executed first and only once before being passed to the outer query.

These types of subqueries are most useful when you need to conduct analysis on a row-level.

### Scalar and non-scalar subqueries

Scalar subqueries are queries that only return a single value. More specifically, this means if you execute a scalar subquery, it would return one column value of one specific row. Non-scalar subqueries, however, can return single or multiple rows and may contain multiple columns.

You may want to use a scalar subquery if you’re interested in passing only a single-row value into an outer query. This type of subquery can be useful when you’re trying to remove or update a specific row’s value using a <Term id="dml">Data Manipulation Language (DML)</Term> statement.

## Subquery examples

> You may often see subqueries in joins and DML statements. We’ll run through examples for each of those scenarios below.

Reviewer 3: I shy away from "below" and "above" in case we change the layout, and it doesn't work well for accessibility or localization reasons.
```suggestion
You may often see subqueries in joins and DML statements. The following sections contain examples for each scenario.
```

### Subquery in a join

In this example, you want to get the lifetime value per customer using your `raw_orders` and `raw_payments` table. Let’s take a look at how you can do that with a subquery in a join:

```sql
select
 
    orders.user_id, 
    sum(payments.amount) as lifetime_value
 

> from {{ ref(`raw_orders`) }} as orders

Reviewer: want this to be `{{ ref('raw_orders') }}`

left join (

    select

        order_id,
        amount

>     from {{ ref(`raw_payments`) }}

Reviewer: want this to also be `{{ ref('raw_payments') }}`

) all_payments
on orders.id = payments.order_id
group by 1
```

Similar to what you saw in the first example, let’s break down the elements of this query.

| Subquery elements | Example |
|---|---|
| Enclosing parentheses | :white_check_mark: |
| Subquery name | `all_payments` |

> | `SELECT` statement | `select order_id, amount from {{ ref(`raw_payments`) }}` |

Reviewer: same as the first example with the single quote issue

> | Main query it is nested in | `select orders.user_id, sum(payments.amount) as lifetime_value from {{ ref(`raw_orders`) }} as orders...` | 

Reviewer: same as the first example with single quote issue

In this example, the `all_payments` subquery will execute first. you use the data from this query to join on the `raw_orders` table to calculate lifetime value per user. Unlike the first example, the subquery happens in the join statement. Subqueries can happen in `JOIN`, `FROM`, and `WHERE` clauses.

### Subquery in a DML command

You may also see subqueries used in DML commands. As a jogger, DML commands are a series of SQL statements that you can write to access and manipulate row-level data in database objects. Oftentimes, you’ll want to use a query result in a qualifying `WHERE` clause to only delete, update, or manipulate certain rows of data.

> In this example below, you'll attempt to update the status of certain orders based on the payment method used in the `raw_payments` table.

Reviewer 3: I try to avoid directional language to assist people with screen readers and when we eventually translate the docs.
```suggestion
In the following example, you'll attempt to update the status of certain orders based on the payment method used in the `raw_payments` table.
```

```sql
UPDATE raw_orders

> set status = ‘returned’

Reviewer: returned should be `'returned'` (single quote string)

where order_id in (
select order_id 
from raw_payments 

> where payment_method  = ‘bank_transfer’)

Reviewer: bank_transfer should be `'bank_transfer'` (single quote string)

```

## Subquery vs CTE

A subquery is a nested query that can oftentimes be used in place of a CTE. Subqueries have different syntax than CTEs, but often have similar use cases. The content won’t go too deep into CTEs here, but it’ll highlight some of the main differences between CTEs and subqueries below.

| CTE | Subquery |
|---|---|
| Typically more readable since CTEs can be used to give structure to your query | Typically less readable, especially if there are many nested queries |
| Reusable in the same query | Must declare the subquery everytime it is used in a query |
| Allows for recursiveness | Does not allow for recursiveness |
| CTEs must have unique CTE_EXPRESSION_NAMES when used in a query | Subqueries don’t always have to be explicitly named |
| CTEs cannot be used in a `WHERE` clause | Subqueries can be used in a `WHERE` clause |

### Subquery vs CTE example

> To demonstrate the similarities and differences between subqueries and CTEs, it’s often most useful to look at an actual example. Going back to the very first example, let's compare how you would perform that query with a subquery vs a CTE:

Reviewer 3: I think we can remove some words to make the language a little simpler. Also "vs" should be avoided for translations purposes.
```suggestion
The following example demonstrates the similarities and differences between subqueries and CTEs. Using the [first subquery example](#subquery-in-a-join), you can compare how you would perform that query using subquery or a CTE:
```

<Tabs
  defaultValue="subquery"
  values={[
    { label: 'Subquery example', value: 'subquery', },
    {label: 'CTE example', value: 'cte', },
  ]
}>
<TabItem value="subquery">

```sql Subquery example
select customer_id, count(order_id) as cnt_orders
  from (

>         select * from {{ ref(‘orders’) }}

Reviewer: This toggle is so beautiful. orders should be `'orders'` (single quote string)

Reviewer 3: Agreed that this is super effective!

       ) all_orders
group by 1
```
</TabItem>
<TabItem value="cte">

```sql CTE example
with all_orders as (

> select * from {{ ref(‘orders’) }}

Reviewer: orders should be `'orders'`

## Subquery syntax

While there are technically several types of subqueries, the general syntax to build them is the same. A subquery usually consists of the following:

- Enclosing parentheses
- A name
- An actual SELECT statement
- A main query it is nested in via a FROM, WHERE, or JOIN clause

Let’s take this to an example, using the [sample jaffle_shop dataset](https://github.com/dbt-labs/jaffle_shop).

```sql
select customer_id, count(order_id) as cnt_orders
  from (
        select * from {{ ref('orders') }}
       ) all_orders
group by 1
```

Given the elements of subqueries laid out in the beginning, let’s break down this example into its respective parts.

| Subquery elements | Example |
|---|---|
| Enclosing parentheses | :white_check_mark: |
| Subquery name | `all_orders` |
| `SELECT` statement | `select * from {{ ref('orders') }}` | 
| Main query it is nested in | `select customer_id, count(order_id) as cnt_orders from all_orders group by 1` | 

> When this query is actually executed, it will start by running the innermost query first. In this case, it would run `select * from {{ ref(‘orders’) }}` first. Then, it would pass those results to the outer query, which is where you grab the count of orders by `customer_id`.

Reviewer: ```suggestion
When this query is actually executed, it will start by running the innermost query first. In this case, it would run `select * from {{ ref('orders') }}` first. Then, it would pass those results to the outer query, which is where you grab the count of orders by `customer_id`.
```

| Subquery elements | Example |
|---|---|
| Enclosing parentheses | :white_check_mark: |
| Subquery name | `all_payments` |
| `SELECT` statement | `select order_id, amount from {{ ref('raw_payments') }}` |
| Main query it is nested in | `select orders.user_id, sum(payments.amount) as lifetime_value from {{ ref('raw_orders') }} as orders...` | 

In this example, the `all_payments` subquery will execute first. you use the data from this query to join on the `raw_orders` table to calculate lifetime value per user. Unlike the first example, the subquery happens in the join statement. Subqueries can happen in `JOIN`, `FROM`, and `WHERE` clauses.

### Subquery in a DML command

You may also see subqueries used in DML commands. As a jogger, DML commands are a series of SQL statements that you can write to access and manipulate row-level data in database objects. Oftentimes, you’ll want to use a query result in a qualifying `WHERE` clause to only delete, update, or manipulate certain rows of data.

In the following example, you'll attempt to update the status of certain orders based on the payment method used in the `raw_payments` table.

```sql
UPDATE raw_orders

> set status = `returned`

Reviewer: ```suggestion
set status = 'returned'
```

where order_id in (
select order_id 
from raw_payments 

> where payment_method  = `bank_transfer`)

Reviewer: ```suggestion
where payment_method  = 'bank_transfer')
```

Subqueries are usually contrasted with <Term id="cte">Common Table Expressions (CTEs)</Term> as they have similar use cases. Unlike CTEs, which are usually separate `SELECT` statements within a query, subqueries are usually `SELECT` statements nested within a `JOIN`, `FROM`, or `WHERE` statement in a query.

To be honest, we rarely write subqueries here at dbt Labs since we prefer to use CTEs. We find that CTEs, in general, support better query readability, organization, and debugging. However, subqueries are a foundational concept in SQL and still widely used. We hope you can use this glossary to better understand how to use subqueries and how they differ from CTEs.

## Subquery syntax

While there are technically several types of subqueries, the general syntax to build them is the same. A subquery usually consists of the following:

- Enclosing parentheses
- A name
- An actual SELECT statement
- A main query it is nested in via a FROM, WHERE, or JOIN clause

Let’s take this to an example, using the [sample jaffle_shop dataset](https://github.com/dbt-labs/jaffle_shop).

```sql
select customer_id, count(order_id) as cnt_orders
  from (

>         select * from {{ ref(`orders`) }}

Reviewer: 'orders' should be single quotes here!

| CTE | Subquery |
|---|---|
| Typically more readable since CTEs can be used to give structure to your query | Typically less readable, especially if there are many nested queries |
| Reusable in the same query | Must declare the subquery everytime it is used in a query |
| Allows for recursiveness | Does not allow for recursiveness |
| CTEs must have unique CTE_EXPRESSION_NAMES when used in a query | Subqueries don’t always have to be explicitly named |
| CTEs cannot be used in a `WHERE` clause | Subqueries can be used in a `WHERE` clause |

### Subquery vs CTE example

To demonstrate the similarities and differences between subqueries and CTEs, it’s often most useful to look at an actual example. Going back to the very first example, let's compare how you would perform that query with a subquery vs a CTE:

<Tabs
  defaultValue="subquery"
  values={[
    { label: 'Subquery example', value: 'subquery', },
    {label: 'CTE example', value: 'cte', },
  ]
}>
<TabItem value="subquery">

```sql Subquery example
select customer_id, count(order_id) as cnt_orders
  from (

>         select * from {{ ref(`orders`) }}

Reviewer: single quotes on orders

       ) all_orders
group by 1
```
</TabItem>
<TabItem value="cte">

```sql CTE example
with all_orders as (

> select * from {{ ref(`orders`) }}

Reviewer: single quotes on orders
